src/query/frontend/semantic/symbol_generator.cpp

```diff
@@ -334,7 +334,15 @@ SymbolGenerator::ReturnType SymbolGenerator::Visit(Identifier &ident) {
       }
       type = scope.visiting_edge->IsVariable() ? Symbol::Type::EDGE_LIST : Symbol::Type::EDGE;
     }
-    symbol = GetOrCreateSymbolLocalScope(ident.name_, ident.user_declared_, type);
+    if (scope.in_foreach) {
+      // Since foreach looks at symbols in a global scope it should also check it
+      if ((scope.in_create_node || scope.in_create_edge) && HasSymbol(ident.name_)) {
+        throw RedeclareVariableError(ident.name_);
+      }
+      symbol = GetOrCreateSymbol(ident.name_, ident.user_declared_, type);
+    } else {
+      symbol = GetOrCreateSymbolLocalScope(ident.name_, ident.user_declared_, type);
+    }
```

Comment:
```suggestion
    symbol = GetOrCreateSymbol(ident.name_, ident.user_declared_, type);
```

Plus changing the similar condition above `if ((scope.in_create_node || scope.in_create_edge) && HasSymbolLocalScope(ident.name_))` to `if ((scope.in_create_node || scope.in_create_edge) && HasSymbol(ident.name_))` should have the same behavior, isn't it?

Reply:
It will have the same behavior for FOREACH, I'm not sure why it is OK to do this for every case. before changes, where you propose putting `GetOrCreateSymbol`, there stood `GetOrCreateSymbolLocalScope`. I don't see why this can be replaced, since the LocalScope variant only looks in the local scope and creates a new symbol if it is not there. This is the scope other clauses are working with. For FOREACH I decided to use global scope to solve shadowing.

Reply:
I now see the comment below regarding scopes, I will try it out a bit and see if it is OK to eliminate local scopes, but then we also need to remove those LocalScope function calls elsewhere